File: src/type4me_linux/history.py
from __future__ import annotations

import base64
import contextlib
import csv
import fcntl
import io
import json
import os
import sqlite3
import threading
import uuid
from collections.abc import Iterable, Iterator, Mapping
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import TextIO

from .paths import AppPaths


CSV_HEADER = (
    "id",
    "created_at",
    "duration_seconds",
    "raw_text",
    "processing_mode",
    "processed_text",
    "final_text",
    "status",
    "character_count",
    "asr_provider",
    "asr_model",
)
_SCHEMA_VERSION = 1
_DEFAULT_BUSY_TIMEOUT_MS = 5_000


class HistoryError(RuntimeError):
    """历史数据库无法安全读取或更新。"""


@dataclass(frozen=True)
class CompletedHistoryRecord:
    raw_text: str
    final_text: str
    duration_seconds: float | None = None
    processing_mode: str | None = None
    processed_text: str | None = None
    status: str = "completed"
    character_count: int | None = None
    asr_provider: str | None = None
    asr_model: str | None = None
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    created_at: str | datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass(frozen=True)
class HistoryRecord:
    id: str
    created_at: str
    duration_seconds: float | None
    raw_text: str
    processing_mode: str | None
    processed_text: str | None
    final_text: str
    status: str
    character_count: int | None
    asr_provider: str | None
    asr_model: str | None


@dataclass(frozen=True)
class HistoryPage:
    records: tuple[HistoryRecord, ...]
    next_cursor: str | None


@dataclass(frozen=True)
class HistoryTotals:
    records: int
    characters: int
    duration_seconds: float


@dataclass(frozen=True)
class UsageSummary:
    days: int
    since: str
    providers: Mapping[str, int]
    models: Mapping[str, int]


class HistoryStore:
    """提供受迁移保护的本地识别历史仓库。"""

    _thread_locks_guard = threading.Lock()
    _thread_locks: dict[Path, threading.Lock] = {}

    def __init__(self, paths: AppPaths, *, busy_timeout_ms: int = _DEFAULT_BUSY_TIMEOUT_MS) -> None:
        if isinstance(busy_timeout_ms, bool) or busy_timeout_ms <= 0:
            raise ValueError("SQLite 忙等待超时值必须是正整数")
        self.path = paths.data / "history.sqlite3"
        self._lock_path = paths.state / "history" / "migration.lock"
        self._busy_timeout_ms = int(busy_timeout_ms)
        self.path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        self._lock_path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        self._migrate()

    def insert(self, record: CompletedHistoryRecord) -> HistoryRecord:
        stored = _stored_record(record)
        values = tuple(getattr(stored, name) for name in CSV_HEADER)
        try:
            with self._connect() as connection:
                connection.execute(
                    """
                    INSERT INTO recognition_history (
                        id, created_at, duration_seconds, raw_text, processing_mode,
                        processed_text, final_text, status, character_count,
                        asr_provider, asr_model
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    values,
                )
        except sqlite3.Error as exc:
            raise HistoryError(f"无法写入识别历史：{exc}") from exc
        return stored

    def get(self, record_id: str) -> HistoryRecord | None:
        try:
            with self._connect() as connection:
                row = connection.execute(
                    f"SELECT {', '.join(CSV_HEADER)} FROM recognition_history WHERE id = ?",
                    (record_id,),
                ).fetchone()
        except sqlite3.Error as exc:
            raise HistoryError(f"无法读取识别历史：{exc}") from exc
        return None if row is None else _row_to_record(row)

    def query(
        self,
        *,
        limit: int = 50,
        cursor: str | None = None,
        from_date: str | datetime | None = None,
        to_date: str | datetime | None = None,
    ) -> HistoryPage:
        if isinstance(limit, bool) or not 1 <= limit <= 1_000:
            raise ValueError("历史分页大小必须在 1 到 1000 之间")

        clauses: list[str] = []
        parameters: list[object] = []
        if cursor is not None:
            cursor_created_at, cursor_id = _decode_cursor(cursor)
            clauses.append("(created_at < ? OR (created_at = ? AND id < ?))")
            parameters.extend((cursor_created_at, cursor_created_at, cursor_id))
        if from_date is not None:
            clauses.append("created_at >= ?")
            parameters.append(_timestamp_text(from_date, "起始日期"))
        if to_date is not None:
            clauses.append("created_at < ?")
            parameters.append(_timestamp_text(to_date, "结束日期"))

        where = f" WHERE {' AND '.join(clauses)}" if clauses else ""
        statement = (
            f"SELECT {', '.join(CSV_HEADER)} FROM recognition_history{where} "
            "ORDER BY created_at DESC, id DESC LIMIT ?"
        )
        parameters.append(limit + 1)
        try:
            with self._connect() as connection:
                rows = connection.execute(statement, parameters).fetchall()
        except sqlite3.Error as exc:
            raise HistoryError(f"无法查询识别历史：{exc}") from exc

        has_more = len(rows) > limit
        records = tuple(_row_to_record(row) for row in rows[:limit])
        next_cursor = None
        if has_more and records:
            last = records[-1]
            next_cursor = _encode_cursor(last.created_at, last.id)
        return HistoryPage(records=records, next_cursor=next_cursor)

    def delete(self, record_id: str) -> bool:
        try:
            with self._connect() as connection:
                result = connection.execute(
                    "DELETE FROM recognition_history WHERE id = ?", (record_id,)
                )
                return result.rowcount == 1
        except sqlite3.Error as exc:
            raise HistoryError(f"无法删除识别历史：{exc}") from exc

    def delete_many(self, record_ids: Iterable[str]) -> int:
        ids = tuple(dict.fromkeys(record_ids))
        if not ids:
            return 0
        placeholders = ", ".join("?" for _ in ids)
        try:
            with self._connect() as connection:
                result = connection.execute(
                    f"DELETE FROM recognition_history WHERE id IN ({placeholders})", ids
                )
                return result.rowcount
        except sqlite3.Error as exc:
            raise HistoryError(f"无法批量删除识别历史：{exc}") from exc

    def delete_all(self) -> int:
        try:
            with self._connect() as connection:
                result = connection.execute("DELETE FROM recognition_history")
                return result.rowcount
        except sqlite3.Error as exc:
            raise HistoryError(f"无法清空识别历史：{exc}") from exc

    def totals(self) -> HistoryTotals:
        try:
            with self._connect() as connection:
                row = connection.execute(
                    """
                    SELECT COUNT(*), COALESCE(SUM(character_count), 0),
                           COALESCE(SUM(duration_seconds), 0.0)
                    FROM recognition_history
                    """
                ).fetchone()
        except sqlite3.Error as exc:
            raise HistoryError(f"无法统计识别历史：{exc}") from exc
        assert row is not None
        return HistoryTotals(
            records=int(row[0]), characters=int(row[1]), duration_seconds=float(row[2])
        )

    def usage_summary(self, days: int, *, now: datetime | None = None) -> UsageSummary:
        if days not in (1, 7, 30):
            raise ValueError("用量统计窗口只能是 1、7 或 30 天")
        current = _aware_utc(now or datetime.now(UTC), "当前时间")
        since = _timestamp_text(current - timedelta(days=days), "统计起始时间")
        try:
            with self._connect() as connection:
                providers = _usage_counts(connection, "asr_provider", since)
                models = _usage_counts(connection, "asr_model", since)
        except sqlite3.Error as exc:
            raise HistoryError(f"无法统计识别用量：{exc}") from exc
        return UsageSummary(days=days, since=since, providers=providers, models=models)

    def usage_summaries(self, *, now: datetime | None = None) -> tuple[UsageSummary, ...]:
        return tuple(self.usage_summary(days, now=now) for days in (1, 7, 30))

    def export_csv(
        self,
        destination: Path | TextIO,
        *,
        from_date: str | datetime | None = None,
        to_date: str | datetime | None = None,
    ) -> int:
        records = self._records_for_export(from_date=from_date, to_date=to_date)
        if hasattr(destination, "write"):
            _write_csv(destination, records)  # type: ignore[arg-type]
        else:
            path = Path(destination)
            path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
            try:
                with path.open("w", encoding="utf-8", newline="") as handle:
                    _write_csv(handle, records)
            except (OSError, UnicodeError) as exc:
                raise HistoryError(f"无法导出识别历史：{exc}") from exc
        return len(records)

    def export_csv_text(
        self,
        *,
        from_date: str | datetime | None = None,
        to_date: str | datetime | None = None,
    ) -> str:
        output = io.StringIO(newline="")
        self.export_csv(output, from_date=from_date, to_date=to_date)
        return output.getvalue()

    def _records_for_export(
        self,
        *,
        from_date: str | datetime | None,
        to_date: str | datetime | None,
    ) -> tuple[HistoryRecord, ...]:
        clauses: list[str] = []
        parameters: list[str] = []
        if from_date is not None:
            clauses.append("created_at >= ?")
            parameters.append(_timestamp_text(from_date, "起始日期"))
        if to_date is not None:
            clauses.append("created_at < ?")
            parameters.append(_timestamp_text(to_date, "结束日期"))
        where = f" WHERE {' AND '.join(clauses)}" if clauses else ""
        try:
            with self._connect() as connection:
                rows = connection.execute(
                    f"SELECT {', '.join(CSV_HEADER)} FROM recognition_history{where} "
                    "ORDER BY created_at DESC, id DESC",
                    parameters,
                ).fetchall()
        except sqlite3.Error as exc:
            raise HistoryError(f"无法读取导出历史：{exc}") from exc
        return tuple(_row_to_record(row) for row in rows)

    def _migrate(self) -> None:
        with self._migration_lock():
            try:
                with self._connect() as connection:
                    connection.execute("BEGIN IMMEDIATE")
                    version = int(connection.execute("PRAGMA user_version").fetchone()[0])
                    if version > _SCHEMA_VERSION:
                        raise HistoryError(f"历史数据库版本过新：{version}")
                    if version < 1:
                        connection.execute(
                            """
                            CREATE TABLE IF NOT EXISTS recognition_history (
                                id TEXT PRIMARY KEY,
                                created_at TEXT NOT NULL,
                                duration_seconds REAL,
                                raw_text TEXT NOT NULL,
                                processing_mode TEXT,
                                processed_text TEXT,
                                final_text TEXT NOT NULL,
                                status TEXT NOT NULL,
                                character_count INTEGER,
                                asr_provider TEXT,
                                asr_model TEXT
                            )
                            """
                        )
                        connection.execute("PRAGMA user_version = 1")
                    _verify_schema(connection)
                    connection.commit()
                os.chmod(self.path, 0o600)
            except HistoryError:
                raise
            except (OSError, sqlite3.Error) as exc:
                raise HistoryError(f"无法迁移历史数据库：{exc}") from exc

    @contextlib.contextmanager
    def _migration_lock(self) -> Iterator[None]:
        resolved = self._lock_path.resolve()
        with self._thread_locks_guard:
            thread_lock = self._thread_locks.setdefault(resolved, threading.Lock())
        with thread_lock:
            try:
                descriptor = os.open(self._lock_path, os.O_RDWR | os.O_CREAT, 0o600)
            except OSError as exc:
                raise HistoryError(f"无法打开历史迁移锁：{exc}") from exc
            try:
                fcntl.flock(descriptor, fcntl.LOCK_EX)
                yield
            except OSError as exc:
                raise HistoryError(f"无法锁定历史数据库迁移：{exc}") from exc
            finally:
                with contextlib.suppress(OSError):
                    fcntl.flock(descriptor, fcntl.LOCK_UN)
                os.close(descriptor)

    @contextlib.contextmanager
    def _connect(self) -> Iterator[sqlite3.Connection]:
        connection: sqlite3.Connection | None = None
        try:
            connection = sqlite3.connect(
                self.path,
                timeout=self._busy_timeout_ms / 1_000,
                isolation_level=None,
            )
            connection.row_factory = sqlite3.Row
            connection.execute(f"PRAGMA busy_timeout = {self._busy_timeout_ms}")
            connection.execute("PRAGMA journal_mode = WAL")
            yield connection
            if connection.in_transaction:
                connection.commit()
        except Exception:
            if connection is not None and connection.in_transaction:
                connection.rollback()
            raise
        finally:
            if connection is not None:
                connection.close()


def _stored_record(record: CompletedHistoryRecord) -> HistoryRecord:
    if not isinstance(record, CompletedHistoryRecord):
        raise TypeError("record 必须是 CompletedHistoryRecord")
    record_id = _required_text(record.id, "历史 ID")
    raw_text = _required_string(record.raw_text, "原始文本")
    final_text = _required_string(record.final_text, "最终文本")
    status = _required_text(record.status, "状态")
    duration = record.duration_seconds
    if duration is not None and (isinstance(duration, bool) or duration < 0):
        raise ValueError("识别时长不能为负数")
    characters = len(final_text) if record.character_count is None else record.character_count
    if isinstance(characters, bool) or characters < 0:
        raise ValueError("字符数不能为负数")
    return HistoryRecord(
        id=record_id,
        created_at=_timestamp_text(record.created_at, "创建时间"),
        duration_seconds=None if duration is None else float(duration),
        raw_text=raw_text,
        processing_mode=_optional_string(record.processing_mode, "处理模式"),
        processed_text=_optional_string(record.processed_text, "处理后文本"),
        final_text=final_text,
        status=status,
        character_count=int(characters),
        asr_provider=_optional_string(record.asr_provider, "ASR 提供方"),
        asr_model=_optional_string(record.asr_model, "ASR 模型"),
    )


def _required_string(value: object, label: str) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{label}必须是字符串")
    return value


def _required_text(value: object, label: str) -> str:
    text = _required_string(value, label)
    if not text:
        raise ValueError(f"{label}不能为空")
    return text


def _optional_string(value: object, label: str) -> str | None:
    if value is None:
        return None
    return _required_string(value, label)


def _timestamp_text(value: str | datetime, label: str) -> str:
    if isinstance(value, datetime):
        parsed = _aware_utc(value, label)
    elif isinstance(value, str):
        try:
            parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        except ValueError as exc:
            raise ValueError(f"{label}不是有效的 RFC 3339 时间") from exc
        parsed = _aware_utc(parsed, label)
    else:
        raise TypeError(f"{label}必须是 datetime 或 RFC 3339 字符串")
    return parsed.isoformat(timespec="microseconds").replace("+00:00", "Z")


def _aware_utc(value: datetime, label: str) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{label}必须包含时区")
    return value.astimezone(UTC)


def _row_to_record(row: sqlite3.Row) -> HistoryRecord:
    return HistoryRecord(**{name: row[name] for name in CSV_HEADER})


def _encode_cursor(created_at: str, record_id: str) -> str:
    payload = json.dumps([created_at, record_id], ensure_ascii=True, separators=(",", ":"))
    return base64.urlsafe_b64encode(payload.encode("ascii")).decode("ascii").rstrip("=")


def _decode_cursor(cursor: str) -> tuple[str, str]:
    if not isinstance(cursor, str) or not cursor:
        raise HistoryError("历史分页游标无效")
    try:
        padded = cursor + "=" * (-len(cursor) % 4)
        raw = base64.b64decode(padded, altchars=b"-_", validate=True)
        payload = json.loads(raw.decode("ascii"))
        if not isinstance(payload, list) or len(payload) != 2:
            raise ValueError
        created_at = _timestamp_text(payload[0], "游标时间")
        record_id = _required_text(payload[1], "游标 ID")
    except (ValueError, TypeError, UnicodeError, json.JSONDecodeError) as exc:
        raise HistoryError("历史分页游标无效") from exc
    return created_at, record_id


def _usage_counts(connection: sqlite3.Connection, column: str, since: str) -> dict[str, int]:
    if column not in {"asr_provider", "asr_model"}:
        raise AssertionError("不支持的用量统计字段")
    rows = connection.execute(
        f"""
        SELECT {column}, COUNT(*) AS uses
        FROM recognition_history
        WHERE created_at >= ? AND {column} IS NOT NULL AND {column} != ''
        GROUP BY {column}
        ORDER BY uses DESC, {column} ASC
        """,
        (since,),
    ).fetchall()
    return {str(row[0]): int(row[1]) for row in rows}


def _write_csv(handle: TextIO, records: tuple[HistoryRecord, ...]) -> None:
    writer = csv.writer(handle, dialect="excel", lineterminator="\r\n")
    writer.writerow(CSV_HEADER)
    for record in records:
        writer.writerow(tuple(getattr(record, name) for name in CSV_HEADER))


def _verify_schema(connection: sqlite3.Connection) -> None:
    expected = (
        ("id", "TEXT", 0, 1),
        ("created_at", "TEXT", 1, 0),
        ("duration_seconds", "REAL", 0, 0),
        ("raw_text", "TEXT", 1, 0),
        ("processing_mode", "TEXT", 0, 0),
        ("processed_text", "TEXT", 0, 0),
        ("final_text", "TEXT", 1, 0),
        ("status", "TEXT", 1, 0),
        ("character_count", "INTEGER", 0, 0),
        ("asr_provider", "TEXT", 0, 0),
        ("asr_model", "TEXT", 0, 0),
    )
    rows = connection.execute("PRAGMA table_info(recognition_history)").fetchall()
    actual = tuple((str(row[1]), str(row[2]).upper(), int(row[3]), int(row[5])) for row in rows)
    if actual != expected:
        raise HistoryError("识别历史表结构与当前版本不兼容")

File: rust/src/history.rs
//! Recognition history persisted in SQLite with the Python-compatible
//! `recognition_history` schema, WAL journaling and 0600 permissions.

use std::path::{Path, PathBuf};

use rusqlite::{params, Connection, OptionalExtension};
use serde::Serialize;
use time::format_description::well_known::Rfc3339;
use time::{Duration, OffsetDateTime};

use crate::coordinator::HistoryEntry;

const SCHEMA_VERSION: i64 = 1;
const BUSY_TIMEOUT_MS: i64 = 5_000;

#[derive(Debug, thiserror::Error)]
pub enum HistoryError {
    #[error("history database error: {0}")]
    Database(String),
    #[error("history timestamp invalid: {0}")]
    Timestamp(String),
    #[error("history cursor invalid")]
    Cursor,
    #[error(transparent)]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct HistoryRecord {
    pub id: String,
    pub created_at: String,
    pub duration_seconds: Option<f64>,
    pub raw_text: String,
    pub processing_mode: Option<String>,
    pub processed_text: Option<String>,
    pub final_text: String,
    pub status: String,
    pub character_count: i64,
    pub asr_provider: Option<String>,
    pub asr_model: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct HistoryPage {
    pub records: Vec<HistoryRecord>,
    pub next_cursor: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct HistoryTotals {
    pub records: i64,
    pub characters: i64,
    pub duration_seconds: f64,
}

pub struct HistoryStore {
    path: PathBuf,
}

impl HistoryStore {
    pub fn open(path: PathBuf) -> Result<Self, HistoryError> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let store = Self { path };
        store.migrate()?;
        Ok(store)
    }

    fn connect(&self) -> Result<Connection, HistoryError> {
        let connection = Connection::open(&self.path)
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        connection
            .pragma_update(None, "busy_timeout", BUSY_TIMEOUT_MS)
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        connection
            .pragma_update(None, "journal_mode", "WAL")
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        Ok(connection)
    }

    fn migrate(&self) -> Result<(), HistoryError> {
        let connection = self.connect()?;
        connection
            .execute_batch("BEGIN IMMEDIATE")
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        let version: i64 = connection
            .pragma_query_value(None, "user_version", |row| row.get(0))
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        if version > SCHEMA_VERSION {
            return Err(HistoryError::Database(format!(
                "history database version too new: {version}"
            )));
        }
        if version < 1 {
            connection
                .execute_batch(
                    "CREATE TABLE IF NOT EXISTS recognition_history (
                        id TEXT PRIMARY KEY,
                        created_at TEXT NOT NULL,
                        duration_seconds REAL,
                        raw_text TEXT NOT NULL,
                        processing_mode TEXT,
                        processed_text TEXT,
                        final_text TEXT NOT NULL,
                        status TEXT NOT NULL,
                        character_count INTEGER,
                        asr_provider TEXT,
                        asr_model TEXT
                    );
                    PRAGMA user_version = 1;",
                )
                .map_err(|error| HistoryError::Database(error.to_string()))?;
        }
        connection
            .execute_batch("COMMIT")
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let permissions = std::fs::Permissions::from_mode(0o600);
            std::fs::set_permissions(&self.path, permissions)?;
        }
        Ok(())
    }

    pub fn insert(&self, entry: &HistoryEntry, backend: &str) -> Result<HistoryRecord, HistoryError> {
        let id = new_record_id();
        let created_at = timestamp_now()?;
        let connection = self.connect()?;
        connection
            .execute(
                "INSERT INTO recognition_history (
                    id, created_at, duration_seconds, raw_text, processing_mode,
                    processed_text, final_text, status, character_count,
                    asr_provider, asr_model
                ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
                params![
                    id,
                    created_at,
                    None::<f64>,
                    entry.raw_text,
                    entry.processing_mode,
                    entry.processed_text,
                    entry.final_text,
                    entry.status,
                    entry.final_text.chars().count() as i64,
                    backend,
                    None::<String>,
                ],
            )
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        Ok(HistoryRecord {
            id,
            created_at,
            duration_seconds: None,
            raw_text: entry.raw_text.clone(),
            processing_mode: Some(entry.processing_mode.clone()),
            processed_text: entry.processed_text.clone(),
            final_text: entry.final_text.clone(),
            status: entry.status.clone(),
            character_count: entry.final_text.chars().count() as i64,
            asr_provider: Some(backend.to_owned()),
            asr_model: None,
        })
    }

    pub fn query(
        &self,
        limit: i64,
        cursor: Option<&str>,
    ) -> Result<HistoryPage, HistoryError> {
        if !(1..=1_000).contains(&limit) {
            return Err(HistoryError::Database(format!(
                "page size must be between 1 and 1000, got {limit}"
            )));
        }
        let connection = self.connect()?;
        let mut statement;
        let rows = if let Some(cursor) = cursor {
            let (created_at, record_id) = decode_cursor(cursor)?;
            statement = connection
                .prepare(
                    "SELECT id, created_at, duration_seconds, raw_text, processing_mode,
                            processed_text, final_text, status, character_count,
                            asr_provider, asr_model
                     FROM recognition_history
                     WHERE (created_at < ?1 OR (created_at = ?1 AND id < ?2))
                     ORDER BY created_at DESC, id DESC LIMIT ?3",
                )
                .map_err(|error| HistoryError::Database(error.to_string()))?;
            statement
                .query_map(params![created_at, record_id, limit + 1], record_from_row)
                .map_err(|error| HistoryError::Database(error.to_string()))?
        } else {
            statement = connection
                .prepare(
                    "SELECT id, created_at, duration_seconds, raw_text, processing_mode,
                            processed_text, final_text, status, character_count,
                            asr_provider, asr_model
                     FROM recognition_history
                     ORDER BY created_at DESC, id DESC LIMIT ?1",
                )
                .map_err(|error| HistoryError::Database(error.to_string()))?;
            statement
                .query_map(params![limit + 1], record_from_row)
                .map_err(|error| HistoryError::Database(error.to_string()))?
        };
        let mut records = Vec::new();
        for row in rows {
            records.push(row.map_err(|error| HistoryError::Database(error.to_string()))?);
        }
        let has_more = records.len() as i64 > limit;
        records.truncate(limit as usize);
        let next_cursor = if has_more {
            records.last().map(|last| encode_cursor(&last.created_at, &last.id))
        } else {
            None
        };
        Ok(HistoryPage {
            records,
            next_cursor,
        })
    }

    pub fn delete(&self, ids: &[String]) -> Result<i64, HistoryError> {
        if ids.is_empty() {
            return Ok(0);
        }
        let connection = self.connect()?;
        let placeholders = vec!["?"; ids.len()].join(", ");
        let deleted = connection
            .execute(
                &format!("DELETE FROM recognition_history WHERE id IN ({placeholders})"),
                rusqlite::params_from_iter(ids),
            )
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        Ok(deleted as i64)
    }

    pub fn delete_all(&self) -> Result<i64, HistoryError> {
        let connection = self.connect()?;
        let deleted = connection
            .execute("DELETE FROM recognition_history", [])
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        Ok(deleted as i64)
    }

    pub fn totals(&self) -> Result<HistoryTotals, HistoryError> {
        let connection = self.connect()?;
        connection
            .query_row(
                "SELECT COUNT(*), COALESCE(SUM(character_count), 0),
                        COALESCE(SUM(duration_seconds), 0.0)
                 FROM recognition_history",
                [],
                |row| {
                    Ok(HistoryTotals {
                        records: row.get(0)?,
                        characters: row.get(1)?,
                        duration_seconds: row.get(2)?,
                    })
                },
            )
            .map_err(|error| HistoryError::Database(error.to_string()))
    }

    pub fn usage(&self, days: i64) -> Result<serde_json::Value, HistoryError> {
        if !matches!(days, 1 | 7 | 30) {
            return Err(HistoryError::Database(format!(
                "usage window must be 1, 7 or 30 days, got {days}"
            )));
        }
        let since = (OffsetDateTime::now_utc() - Duration::days(days))
            .format(&Rfc3339)
            .map_err(|error| HistoryError::Timestamp(error.to_string()))?;
        let connection = self.connect()?;
        let mut providers: std::collections::BTreeMap<String, i64> =
            std::collections::BTreeMap::new();
        let mut models: std::collections::BTreeMap<String, i64> =
            std::collections::BTreeMap::new();
        let mut statement = connection
            .prepare(
                "SELECT COALESCE(asr_provider, ''), COUNT(*) FROM recognition_history
                 WHERE created_at >= ?1 GROUP BY 1",
            )
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        let rows = statement
            .query_map(params![since], |row| Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?)))
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        for row in rows {
            let (key, count) = row.map_err(|error| HistoryError::Database(error.to_string()))?;
            if !key.is_empty() {
                providers.insert(key, count);
            }
        }
        let mut statement = connection
            .prepare(
                "SELECT COALESCE(asr_model, ''), COUNT(*) FROM recognition_history
                 WHERE created_at >= ?1 GROUP BY 1",
            )
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        let rows = statement
            .query_map(params![since], |row| Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?)))
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        for row in rows {
            let (key, count) = row.map_err(|error| HistoryError::Database(error.to_string()))?;
            if !key.is_empty() {
                models.insert(key, count);
            }
        }
        Ok(serde_json::json!({
            "days": days,
            "since": since,
            "providers": providers,
            "models": models,
        }))
    }

    pub fn export_csv(&self, destination: &Path) -> Result<i64, HistoryError> {
        let connection = self.connect()?;
        let mut statement = connection
            .prepare(
                "SELECT id, created_at, duration_seconds, raw_text, processing_mode,
                        processed_text, final_text, status, character_count,
                        asr_provider, asr_model
                 FROM recognition_history ORDER BY created_at DESC, id DESC",
            )
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        let records: Vec<HistoryRecord> = statement
            .query_map([], record_from_row)
            .map_err(|error| HistoryError::Database(error.to_string()))?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        let mut csv = String::from(
            "id,created_at,duration_seconds,raw_text,processing_mode,processed_text,final_text,status,character_count,asr_provider,asr_model\n",
        );
        for record in &records {
            let fields = [
                &record.id,
                &record.created_at,
                &record
                    .duration_seconds
                    .map(|value| value.to_string())
                    .unwrap_or_default(),
                &record.raw_text,
                record.processing_mode.as_deref().unwrap_or_default(),
                record.processed_text.as_deref().unwrap_or_default(),
                &record.final_text,
                &record.status,
                &record.character_count.to_string(),
                record.asr_provider.as_deref().unwrap_or_default(),
                record.asr_model.as_deref().unwrap_or_default(),
            ];
            let escaped: Vec<String> = fields
                .iter()
                .map(|field| {
                    if field.contains(',') || field.contains('"') || field.contains('\n') {
                        format!("\"{}\"", field.replace('"', "\"\""))
                    } else {
                        (*field).to_string()
                    }
                })
                .collect();
            csv.push_str(&escaped.join(","));
            csv.push('\n');
        }
        if let Some(parent) = destination.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::write(destination, csv)?;
        Ok(records.len() as i64)
    }

    pub fn get(&self, record_id: &str) -> Result<Option<HistoryRecord>, HistoryError> {
        let connection = self.connect()?;
        let record = connection
            .query_row(
                "SELECT id, created_at, duration_seconds, raw_text, processing_mode,
                        processed_text, final_text, status, character_count,
                        asr_provider, asr_model
                 FROM recognition_history WHERE id = ?1",
                params![record_id],
                record_from_row,
            )
            .optional()
            .map_err(|error| HistoryError::Database(error.to_string()))?;
        Ok(record)
    }
}

fn record_from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<HistoryRecord> {
    Ok(HistoryRecord {
        id: row.get(0)?,
        created_at: row.get(1)?,
        duration_seconds: row.get(2)?,
        raw_text: row.get(3)?,
        processing_mode: row.get(4)?,
        processed_text: row.get(5)?,
        final_text: row.get(6)?,
        status: row.get(7)?,
        character_count: row.get(8)?,
        asr_provider: row.get(9)?,
        asr_model: row.get(10)?,
    })
}

fn timestamp_now() -> Result<String, HistoryError> {
    OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .map_err(|error| HistoryError::Timestamp(error.to_string()))
}

fn new_record_id() -> String {
    let mut bytes = [0_u8; 16];
    getrandom::getrandom(&mut bytes).expect("OS random source available");
    bytes[6] = (bytes[6] & 0x0f) | 0x40;
    bytes[8] = (bytes[8] & 0x3f) | 0x80;
    let hex: String = bytes.iter().map(|byte| format!("{byte:02x}")).collect();
    format!(
        "{}-{}-{}-{}-{}",
        &hex[0..8],
        &hex[8..12],
        &hex[12..16],
        &hex[16..20],
        &hex[20..32]
    )
}

fn encode_cursor(created_at: &str, record_id: &str) -> String {
    use base64::engine::general_purpose::URL_SAFE_NO_PAD;
    use base64::Engine;
    let payload = serde_json::json!([created_at, record_id]).to_string();
    URL_SAFE_NO_PAD.encode(payload.as_bytes())
}

fn decode_cursor(cursor: &str) -> Result<(String, String), HistoryError> {
    use base64::engine::general_purpose::URL_SAFE_NO_PAD;
    use base64::Engine;
    let bytes = URL_SAFE_NO_PAD
        .decode(cursor.as_bytes())
        .map_err(|_| HistoryError::Cursor)?;
    let value: serde_json::Value =
        serde_json::from_slice(&bytes).map_err(|_| HistoryError::Cursor)?;
    let created_at = value
        .get(0)
        .and_then(serde_json::Value::as_str)
        .ok_or(HistoryError::Cursor)?
        .to_owned();
    let record_id = value
        .get(1)
        .and_then(serde_json::Value::as_str)
        .ok_or(HistoryError::Cursor)?
        .to_owned();
    Ok((created_at, record_id))
}
